text/0000-code-literals.md

```diff
@@ -0,0 +1,511 @@
+
+```rust
+let code = hr#"
+    Text with no final newline
+    -"#;
+```
+
+Aside from this `-` modifier, only whitespace may appear on the final
+line prior to the closing quote.
+
+Together, these rules ensure that every possible string can be represented
+in a single canonical way, while allowing the indentation of the string
+as a whole to be changed freely.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+An `h` modifier may be added to the prefix of the following string
+literal types:
+
+- String literals `h"`
+- Raw string literals `hr#"`
+- Byte string literals `hb"`
+- Raw byte string literals `hbr#"`
+
+The `h` modifier will appear before all characters in the prefix.
+
+The value of a string literal with the `h` modifier will be determined
+using the following steps:
+
+1.  Measure the whitespace indenting the closing quote. If a
+    non-whitespace character (other than a single `-`) exists before
+    the closing quote on the same line, then issue a compiler error.
+2.  Take the lines *between* (but not including) the opening and
+    closing quotes exactly as written.
+3.  Remove exactly the measured whitespace from each non-empty line.
+    If this cannot be done, then issue a compiler error. The
+    whitespace must match down to the exact character sequence.
+4.  If a `-` character was present immediately prior to the closing
+    quote, then remove the final newline.
+5.  Interpret any escape sequences and apply any pre-processing as
+    usual for the string literal type without an `h` modifier.
+    For example, newlines in the file are always treated as `\n`
+    even if the file is encoded with `\r\n` newlines.
+
+Here are some edge case examples:
+
+```rust
+    // Empty string with language hint
+    assert_eq!(h"foo
+        ", "");
+
+    // Newline
+    assert_eq!(h"
+
+        ", "\n");
+
+    // No terminating newline
+    assert_eq!(h"
+        bar
+        -", "bar");
+
+    // Terminating newline
+    assert_eq!(h"
+        bar
+        ", "bar\n");
+
+    // Preserved indent
+    assert_eq!(hr#"
+    if a:
+        print(42)
+    "#, "if a:\n    print(42)\n");
+
+    // Relative indent
+    assert_eq!(hr#"
+            if a:
+                print(42)
+            "#, "if a:\n    print(42)\n");
+
+    // Relative to closing quote
+    assert_eq!(hr#"
+
+
+            if a:
+                print(42)
+    "#, "\n\n        if a:\n            print(42)\n");
+
+    // Interactions with escaping rules
+    assert_eq!(h"
+        \"\
+            foo\n
+            bar
+    \t
+    ", "    \"foo\n\n        bar\n\t\n");
+```
+
+Any text between the opening quote and the first newline is
+preserved within the AST, but is otherwise unused.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The main drawback is increased complexity of the language:
+
+1.  It adds a four new types of string literals given all
+    the combinations.
+
+# Rationale and alternatives
```

Comment:
Another alternative [I have mentioned on zulip](https://rust-lang.zulipchat.com/#narrow/stream/213817-t-lang/topic/Review.20of.20code.20string.20literals.20proposal/near/371587454):

Improve `include!` handling (when passed as literals to macros? in editors?) instead to make it more ergonomic to outline other-language code rather than inlining.

Pros:

- works better with simple tools that don't handle *nested* languages well
- establishes a new indent context, i.e. doesn't need to be adjusted with surrounding code which in my experience can be error-prone if the editor's indentation handling is imperfect. Examples of confusions:
  - inside comments
  - inside macros
  - inside doc comments
  - when current indentation is inconsistent with configured rules
  - when copy-pasting into a differently indented context
- generally avoids stacking complexity

  ```rust
  some_proc_macro!{
     mod m {
        /// This is an example with nesting and several levels of indentation and whitespaces
        ///
        /// ```rust
        /// let p = h"python
        ///      def py():
        ///          a = '''Lorem ipsum dolor sit amet,
        ///          consectetur adipiscing elit,
        ///          sed do eiusmod tempor incididunt
        ///          ut labore et dolore magna aliqua.'''
        ///          print(a) 
        ///      ";
        /// ```
        ///
        fn nesting_fun() {}
     }
  }
  ```

Cons:

- requires editor support if you want to view or even edit the included file in the context of its parent instead of opening a new view. But showing an overlay might be less complex than all the text nesting
- context of substitutions may be harder to see


Since this is motivated by making things easier for rustfmt I recommend contacting the maintainers of other tools (syntax highlighters, editors, IDEs, ...) to see if this change helps or adds complexity for them.

Reply:
I don't consider this an alternative. Requiring powerful editor support to even use the feature makes it a no-go, and having to store things in separate files is a maintenance burden that's worse than the current situation, since it requires coming up with a naming scheme for those files that makes sense, makes it harder to resolve merge conflicts since tools like `git` will never understand this "magic include", and is way more complicated than what is proposed in this RFC.

The advantages you list I also consider to be problems with your approach. You say it works better with simple tools, but the opposite is true: you end up with something unworkable without powerful editor features. In contrast this RFC doesn't require any editor features at all to be an improvement over the status quo. Any support for nested language is an optional extra that doesn't affect the core functionality.

Your example of "stacking complexity" seems very straightforward tbh. Infinitely better than having to go to a spearate file.

> Since this is motivated by making things easier for rustfmt I recommend contacting the maintainers of other tools (syntax highlighters, editors, IDEs, ...) to see if this change helps or adds complexity for them.

It by definition does not add any complexity for tools other than rustfmt, since the only required change as a result of this RFC is allowing a new prefix letter (`h` proposed here) and tools must already support that. Beside that, anything that is valid to do with a raw string literal is also valid to do with an `h` raw string literal.

Reply:
> It by definition does not add any complexity for tools other than rustfmt, since the only required change as a result of this RFC is allowing a new prefix letter (`h` proposed here) and tools must already support that. Beside that, anything that is valid to do with a raw string literal is also valid to do with an `h` raw string literal.

Anything that adds syntax complicates `syn` and any other tools that use it or otherwise parse rust code. I can't imagine that it would ever be safe to just assume that any string prefix acts like a regular string literal, since raw strings already violate that, hence individual new letters have to be added to anything that parses rust code, including syntax highlighters (although the backup behavior is usually good enough for these).

Reply:
> You end up with something unworkable without powerful editor features.

How? Even many simple editors at least have tabs, panes or similar UI elements to view more than one file at a time. 

At its most primitive you rely on your window manager and file browser to open multiple files at the same time in separate windows and show them side by side.

> Any support for nested language is an optional extra that doesn't affect the core functionality.

A simple editor can have primitive syntax-highlighting that will work with separate files based on file extensions but won't work with inlined content. So this RFC makes things worse for simple editors

> makes it harder to resolve merge conflicts since tools like git will never understand this "magic include"

I don't see how it would make things more difficult for git? If anything it makes diffs simple due to fewer whitespace adjustments.

> Requiring powerful editor support to even use the feature makes it a no-go,

Where did I said that a powerful editor would be *required*? Rather I'm suggesting
a) improve powerful editors
b) keep things simple for simple editors

This covers both.

> Your example of "stacking complexity" seems very straightforward tbh. Infinitely better than having to go to a separate file.

What is straight-forward about it? If you actually have to edit, indent, copy-paste, syntax-highlight or auto-complete that there are lots of pitfalls.
Note the outer macro which tends to make things more difficult for tools because at that point point they might not even know anymore whether they're dealing with rust or just things that happen to tokenize like rust.
And it's rust -> macro -> markdown -> codeblock (with language annotation) -> multiline string (with another language annotation).
These languages could be configured to have different indent rules!

Reply:
I don't think improving support for `include!` like things to be a _negative_ (proc-macro-include RFC, proc-macro-expand feature would both be great to have), but it's a feature for different usecases than this. This RFC improves support for things that _people are already doing_. Even if we had better forms of `include!` I would not pull out 3 lines of SQL to a separate file just to get syntax highlighting, I would simply do what we do currently: use the existing literal strings and fight with rustfmt every time the surrounding code changes.

Reply:
> Even if we had better forms of include! I would not pull out 3 lines of SQL to a separate file just to get syntax highlighting

Why not? What do you imagine about it still being inconvenient about that? Whatever that is, I'm saying should be improved.

Reply:
Improving that requires no changes on Rust's side. That's entirely about the editor experience, and I don't see any easy features that would integrate into my setup to improve it.

Reply:
Well, let's consider non-easy ones then. Perhaps a LSP extension to indicate inlined content would make sense.

Reply:
> How? Even many simple editors at least have tabs, panes or similar UI elements to view more than one file at a time.

If we have to view more than one file at a time to use the feature then it's already worse than the status quo.

> A simple editor can have primitive syntax-highlighting that will work with separate files based on file extensions but won't work with inlined content. So this RFC makes things worse for simple editors

No it doesn't. You're completely missing the point that *these strings are already in the file*. No editors currently syntax highlight these strings. This RFC makes it *possible* for advanced editors to support syntax highlighting these strings. Simple editors are unaffected.

> I don't see how it would make things more difficult for git? If anything it makes diffs simple due to fewer whitespace adjustments.

At the moment, these strings are *in the file* and so can be reviewed and have conflicts resolved in-place. By moving them to a separate file you can no longer perform these actions with any context about the surrounding code.

> Where did I said that a powerful editor would be required? Rather I'm suggesting

You suggested moving them to a separate file. To make that at all workable you'd need a powerful editor to allow treating them as though they weren't in a separate file, since having to edit them in a separate file is so much worse than what we already have today.

Reply:
Separation of languages is the norm and should be encouraged.  See the HTML/CSS/JS split that is encouraged instead of having inline script handlers and styles. See template files. See module trees.

You say my approach is a no-go because it makes things more difficult for simple editors. And yet you acknowledge that this RFC will primarily benefit complex editors. While I think my approach would benefit simple editors because they can then work with the outlined language.

> At the moment, these strings are in the file and so can be reviewed and have conflicts resolved in-place. By moving them to a separate file you can no longer perform these actions with any context about the surrounding code.

I assume they'd conventionally still be placed in the same directory and show up in the diffs next to each other.

> To make that at all workable you'd need a powerful editor to allow treating them as though they weren't in a separate file

Not necessarily. E.g. when you have an SQL query `query!(include!("query.psql"), param1="val", param2="val", ...)` then it has an API, like a function call. You edit functions separately and then fix their callsites.
So "jump to definition" + error messages from the `query!` macro about missing arguments would already cover that.

Reply:
> And yet you acknowledge that this RFC will primarily benefit complex editors. 

My expectation is that this RFC will not effect complex editors (in cases where they are not acting as simple editors).

A complex editor that is using heuristics to determine when to apply other-language syntax highlighting to a literal could similarly use those heuristics to determine when to other-language auto-formatting to a literal.

This RFC simply provides support for _auto-indentation_ (but not formatting) of literals for simple editors (and complex editors where their heuristics don't apply) that use rustfmt.

Reply:
> Separation of languages is the norm and should be encouraged. 

That is not my experience. I've almost never seen sql queries pulled out into separate files. Most assembly I've seen is inline. Shader languages are a bit of  a mix,  and I don't have as much familiarity with it, but I don't think it is at all unusual to include shader code inline, especially if it is small. And this feature would be very useful for help text for cli programs. I can't imagine using a separate file for the help comment for every option in my cli that uses clap. 

> See the HTML/CSS/JS split that is encouraged instead of having inline script handlers and styles

But we also have frameworks like react, where html and css are embedded in Javascript. Or svelte where the JS is included in an html template.

Reply:
Shader languages are the one case I can think of where people actually care about "separation of languages", and then it has to do more with the fact that GPU code inherently has a modularity to it, because it is run in passes, and people tend to pull out modules into, well, modules. So you may as well have, e.g.
- code.cpp
- code.hpp
- code.vert
- code.frag

But ofc you may well just encounter something like
- code.cs
- code.hlsl

Depending.

Reply:
> notepad++

Has syntax highlighting.